File: retail_tools/retail_tools/page/label_generator/label_generator.py
"""
Label Generator - Backend API

Provides API endpoints for generating barcode labels with:
- Item lookup with barcode data
- Multiple label formats
- Quantity support
"""

import frappe
from frappe import _


# Label format configurations (width in mm, columns)
LABEL_FORMATS = {
    "small": {"name": "Small (38x25mm)", "columns": 4, "width": 38, "height": 25},
    "medium": {"name": "Medium (50x30mm)", "columns": 3, "width": 50, "height": 30},
    "large": {"name": "Large (70x40mm)", "columns": 2, "width": 70, "height": 40},
}


@frappe.whitelist()
def get_label_formats() -> dict:
    """Get available label formats."""
    return {"ok": True, "formats": LABEL_FORMATS}


@frappe.whitelist()
def get_items_with_stock(warehouse: str = None) -> dict:
    """
    Get all items with stock for label generation.

    Args:
        warehouse: Optional warehouse filter

    Returns:
        dict with items array
    """
    # Get items with stock from Bin
    filters = {}
    if warehouse:
        filters["warehouse"] = warehouse

    bins = frappe.get_all(
        "Bin",
        filters=filters,
        fields=["item_code", "actual_qty", "warehouse"],
        order_by="item_code",
    )

    # Filter to positive stock and aggregate by item
    from collections import defaultdict
    stock_by_item = defaultdict(float)
    for b in bins:
        if b.actual_qty > 0:
            stock_by_item[b.item_code] += b.actual_qty

    if not stock_by_item:
        return {"ok": False, "message": _("No items with stock found")}

    # Get item details
    items = []
    for item_code, qty in stock_by_item.items():
        result = get_item_for_label(item_code)
        if result.get("ok"):
            item = result["item"]
            item["qty"] = int(qty)
            items.append(item)

    return {"ok": True, "items": items, "count": len(items)}


@frappe.whitelist()
def get_items_from_receipt_document(doctype: str, docname: str) -> dict:
    """
    Get items from a Purchase Invoice or Purchase Receipt for label generation.

    Args:
        doctype: Document type (Purchase Invoice or Purchase Receipt)
        docname: Document name

    Returns:
        dict with items array (item_code, item_name, barcode, price, qty)
    """
    # Validate doctype
    allowed_doctypes = ["Purchase Invoice", "Purchase Receipt"]
    if doctype not in allowed_doctypes:
        return {"ok": False, "message": _("Invalid document type")}

    # Validate document exists
    if not frappe.db.exists(doctype, docname):
        return {"ok": False, "message": _("Document not found")}

    # Get child table name
    child_doctype = f"{doctype} Item"

    # Get items from child table
    items_data = frappe.get_all(
        child_doctype,
        filters={"parent": docname},
        fields=["item_code", "qty"],
        order_by="idx",
    )

    if not items_data:
        return {"ok": False, "message": _("No items found in document")}

    # Get item details for each item
    items = []
    for row in items_data:
        result = get_item_for_label(row.item_code)
        if result.get("ok"):
            item = result["item"]
            item["qty"] = int(row.qty)
            items.append(item)

    if not items:
        return {"ok": False, "message": _("No valid items found")}

    return {"ok": True, "items": items, "count": len(items)}


@frappe.whitelist()
def get_item_for_label(item_code: str) -> dict:
    """
    Get item data for label generation.

    Args:
        item_code: Item code to look up

    Returns:
        dict with item data including barcode
    """
    if not item_code:
        return {"ok": False, "message": _("Item code is required")}

    # Check if item exists
    if not frappe.db.exists("Item", item_code):
        return {"ok": False, "message": _("Item not found")}

    # Get item data
    item = frappe.get_doc("Item", item_code)

    # Get barcode
    barcode = None
    if item.barcodes:
        barcode = item.barcodes[0].barcode

    # Get price from default selling price list
    price = 0
    default_price_list = frappe.db.get_single_value("Selling Settings", "selling_price_list")
    if default_price_list:
        price_doc = frappe.db.get_value(
            "Item Price",
            {"item_code": item_code, "price_list": default_price_list, "selling": 1},
            "price_list_rate",
        )
        if price_doc:
            price = price_doc

    return {
        "ok": True,
        "item": {
            "item_code": item.item_code,
            "item_name": item.item_name,
            "barcode": barcode,
            "price": price,
        },
    }


@frappe.whitelist()
def generate_labels_html(items: str, label_format: str = "medium", show_price: int = 1) -> dict:
    """
    Generate HTML for label printing.

    Args:
        items: JSON string of items with quantities [{item_code, qty}, ...]
        label_format: Label format key (small, medium, large)

    Returns:
        dict with HTML content
    """
    import json

    try:
        items_list = json.loads(items) if isinstance(items, str) else items
    except json.JSONDecodeError:
        return {"ok": False, "message": _("Invalid items data")}

    if not items_list:
        return {"ok": False, "message": _("No items provided")}

    # Get format config
    format_config = LABEL_FORMATS.get(label_format, LABEL_FORMATS["medium"])
    columns = format_config["columns"]

    # Collect all labels
    labels = []
    for item_data in items_list:
        item_code = item_data.get("item_code")
        qty = int(item_data.get("qty", 1))

        # Get item details
        result = get_item_for_label(item_code)
        if not result.get("ok"):
            continue

        item = result["item"]

        # Add label for each quantity
        for _ in range(qty):
            labels.append(item)

    if not labels:
        return {"ok": False, "message": _("No valid items found")}

    # Build HTML
    html = _build_labels_html(labels, format_config, int(show_price))

    return {"ok": True, "html": html, "count": len(labels)}


def _build_labels_html(labels: list, format_config: dict, show_price: int = 1) -> str:
    """Build labels HTML grid."""
    columns = format_config["columns"]
    label_width = format_config["width"]
    label_height = format_config["height"]

    # Check if small format (horizontal layout)
    is_small = columns == 4

    # Adjust sizes based on format (fewer columns = larger labels)
    name_size = "11px"
    price_size = "12px"
    code_size = "9px"
    barcode_height = "35px"
    char_limit = 30
    
    if columns == 2:  # Large labels
        name_size = "14px"
        price_size = "16px"
        code_size = "11px"
        barcode_height = "55px"
        char_limit = 40
    elif is_small:  # Small labels - horizontal layout
        name_size = "11px"
        price_size = "11px"
        code_size = "9px"
        barcode_height = "40px"
        char_limit = 18

    labels_html = ""
    for label in labels:
        formatted_price = ""
        if show_price and label.get("price"):
            formatted_price = frappe.format_value(
                label["price"], {"fieldtype": "Currency"}
            )

        barcode_html = ""
        if label.get("barcode"):
            barcode_html = f'<svg class="barcode-svg" data-barcode="{label["barcode"]}"></svg>'
        else:
            barcode_html = f'<div class="label-no-barcode">{_("No barcode")}</div>'

        if is_small:
            # Small labels: barcode on top, info row below (name/code left, price right)
            # If no price, center the info
            price_html = f'<div class="label-price-side">{formatted_price}</div>' if formatted_price else ""
            centered_class = "" if formatted_price else " label-centered"
            labels_html += f"""
            <div class="label-item label-small{centered_class}">
                <div class="label-barcode">{barcode_html}</div>
                <div class="label-info-row">
                    <div class="label-left">
                        <div class="label-name">{label["item_name"][:char_limit]}</div>
                        <div class="label-code">{label["item_code"]}</div>
                    </div>
                    {price_html}
                </div>
            </div>
            """
        else:
            # Vertical layout for medium/large labels
            price_html = f'<div class="label-price">{formatted_price}</div>' if formatted_price else ""
            labels_html += f"""
            <div class="label-item">
                <div class="label-barcode">{barcode_html}</div>
                <div class="label-name">{label["item_name"][:char_limit]}</div>
                <div class="label-code">{label["item_code"]}</div>
                {price_html}
            </div>
            """

    return f"""
    <style>
        .labels-container {{
            font-family: Arial, sans-serif;
        }}
        .labels-grid {{
            display: grid;
            grid-template-columns: repeat({columns}, 1fr);
            gap: 5px;
        }}
        .label-item {{
            border: 1px dashed #ccc;
            padding: 8px;
            text-align: center;
            min-height: {label_height}mm;
            box-sizing: border-box;
            page-break-inside: avoid;
            display: flex;
            flex-direction: column;
            justify-content: center;
        }}
        /* Small label layout: barcode top, info row below */
        .label-small {{
            padding: 4px 8px;
        }}
        .label-small .label-barcode {{
            margin-bottom: 4px;
            text-align: center;
        }}
        .label-small .label-barcode svg {{
            height: {barcode_height};
            max-width: 100%;
        }}
        .label-info-row {{
            display: flex;
            align-items: center;
            justify-content: space-between;
        }}
        .label-small .label-left {{
            flex: 1;
            min-width: 0;
            text-align: left;
        }}
        .label-small .label-name {{
            font-size: {name_size};
            font-weight: 600;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: nowrap;
        }}
        .label-small .label-code {{
            font-size: {code_size};
            color: #666;
        }}
        .label-price-side {{
            font-weight: 700;
            font-size: {price_size};
            color: #333;
            text-align: right;
            white-space: nowrap;
            padding-left: 8px;
        }}
        /* Centered layout when no price */
        .label-centered .label-info-row {{
            justify-content: center;
        }}
        .label-centered .label-left {{
            text-align: center;
        }}
        /* Vertical layout (medium/large) */
        .label-barcode {{
            margin-bottom: 4px;
        }}
        .label-barcode svg {{
            max-width: 100%;
            height: {barcode_height};
        }}
        .label-no-barcode {{
            color: #999;
            font-size: 10px;
            padding: 10px 0;
        }}
        .label-name {{
            font-weight: 600;
            font-size: {name_size};
            line-height: 1.2;
            margin-bottom: 2px;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: nowrap;
        }}
        .label-code {{
            font-size: {code_size};
            color: #666;
            margin-bottom: 2px;
        }}
        .label-price {{
            font-weight: 700;
            font-size: {price_size};
            color: #333;
            margin-top: auto;
        }}
        @media print {{
            .label-item {{
                border: 1px solid #ddd;
            }}
        }}
    </style>
    <div class="labels-container">
        <div class="labels-grid">
            {labels_html}
        </div>
    </div>
    """

File: retail_tools/retail_tools/page/label_generator/label_generator.js
/* global frappe, __ */
/* eslint-disable no-undef */

/**
 * Label Generator Page
 *
 * Generate printable barcode labels with:
 * - Item selection with quantity
 * - Multiple label formats
 * - Print support
 */

frappe.pages["label-generator"].on_page_load = function (wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __("Label Generator"),
        single_column: true,
    });

    page.label_generator = new retail_tools.LabelGenerator(page, wrapper);
};

frappe.provide("retail_tools");

retail_tools.LabelGenerator = class LabelGenerator {
    constructor(page, wrapper) {
        this.page = page;
        this.wrapper = wrapper;
        this.$body = $(this.page.body);
        this.items = [];

        this.make_actions();
        this.make_layout();
    }

    make_actions() {
        this.page.set_primary_action(__("Generate Labels"), () =>
            this.generate_labels()
        );

        this.page.add_inner_button(__("Print"), () => this.print_labels());

        this.page.add_inner_button(__("Clear All"), () => this.clear_items());
    }

    make_layout() {
        this.$body.html(`
      <div class="lg-wrapper">
        <div class="lg-controls">
          <div class="lg-add-item">
            <div class="lg-field" id="lg-item-field"></div>
            <div class="lg-field lg-qty-field" id="lg-qty-field"></div>
            <button class="btn btn-primary lg-add-btn">${__("Add")}</button>
          </div>
          <div class="lg-bulk-row">
            <div class="lg-field" id="lg-warehouse-field"></div>
            <button class="btn btn-default lg-load-stock-btn">
              <i class="fa fa-download"></i> ${__("Load Items with Stock")}
            </button>
          </div>
          <div class="lg-bulk-row">
            <div class="lg-field lg-field-narrow" id="lg-receipt-doctype-field"></div>
            <div class="lg-field" id="lg-receipt-document-field"></div>
            <button class="btn btn-default lg-load-receipt-btn">
              <i class="fa fa-file-text-o"></i> ${__("Load from Document")}
            </button>
          </div>
          <div class="lg-format-row">
            <label>${__("Label Format")}:</label>
            <div class="lg-field" id="lg-format-field"></div>
            <div class="lg-field" id="lg-price-list-field"></div>
            <label class="lg-checkbox">
              <input type="checkbox" id="lg-show-price" checked>
              ${__("Show Price")}
            </label>
          </div>
        </div>
        <div class="lg-items-list">
          <div class="lg-items-header">
            <span class="lg-col-item">${__("Item")}</span>
            <span class="lg-col-barcode">${__("Barcode")}</span>
            <span class="lg-col-qty">${__("Qty")}</span>
            <span class="lg-col-actions"></span>
          </div>
          <div class="lg-items-body"></div>
          <div class="lg-items-empty">
            <p class="text-muted">${__("Add items to generate labels")}</p>
          </div>
        </div>
        <div class="lg-preview-container">
          <div class="lg-preview"></div>
        </div>
      </div>
    `);

        // Create Item field
        this.item_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Link",
                fieldname: "item_code",
                label: __("Item"),
                options: "Item",
                placeholder: __("Select or scan item..."),
            },
            parent: this.$body.find("#lg-item-field"),
            render_input: true,
        });
        this.item_field.refresh();

        // Create Quantity field
        this.qty_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Int",
                fieldname: "qty",
                label: __("Qty"),
                default: 1,
            },
            parent: this.$body.find("#lg-qty-field"),
            render_input: true,
        });
        this.qty_field.set_value(1);
        this.qty_field.refresh();

        // Create Format field
        this.format_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Select",
                fieldname: "label_format",
                label: __("Format"),
                options: [
                    { value: "small", label: __("Small (38x25mm) - 4 cols") },
                    { value: "medium", label: __("Medium (50x30mm) - 3 cols") },
                    { value: "large", label: __("Large (70x40mm) - 2 cols") },
                ],
                default: "medium",
            },
            parent: this.$body.find("#lg-format-field"),
            render_input: true,
        });
        this.format_field.set_value("medium");
        this.format_field.refresh();

        // Create Price List field
        this.price_list_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Link",
                fieldname: "price_list",
                label: __("Price List"),
                options: "Price List",
                placeholder: __("Default"),
                get_query: () => {
                    return { filters: { selling: 1 } };
                },
            },
            parent: this.$body.find("#lg-price-list-field"),
            render_input: true,
        });
        this.price_list_field.refresh();

        // Create Warehouse field
        this.warehouse_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Link",
                fieldname: "warehouse",
                label: __("Warehouse"),
                options: "Warehouse",
                placeholder: __("All warehouses"),
            },
            parent: this.$body.find("#lg-warehouse-field"),
            render_input: true,
        });
        this.warehouse_field.refresh();

        // Create Receipt Document Type field
        this.receipt_doctype_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Select",
                fieldname: "receipt_doctype",
                label: __("Document Type"),
                options: [
                    { value: "Purchase Invoice", label: __("Purchase Invoice") },
                    { value: "Purchase Receipt", label: __("Purchase Receipt") },
                ],
                default: "Purchase Invoice",
            },
            parent: this.$body.find("#lg-receipt-doctype-field"),
            render_input: true,
        });
        this.receipt_doctype_field.set_value("Purchase Invoice");
        this.receipt_doctype_field.refresh();

        // Create Receipt Document field (Dynamic Link)
        this.receipt_document_field = frappe.ui.form.make_control({
            df: {
                fieldtype: "Link",
                fieldname: "receipt_document",
                label: __("Document"),
                options: "Purchase Invoice",
                placeholder: __("Select document..."),
            },
            parent: this.$body.find("#lg-receipt-document-field"),
            render_input: true,
        });
        this.receipt_document_field.refresh();

        // Bind doctype change to update document link options
        this.receipt_doctype_field.$input.on("change", () => {
            const doctype = this.receipt_doctype_field.get_value();
            this.receipt_document_field.df.options = doctype;
            this.receipt_document_field.set_value("");
            this.receipt_document_field.refresh();
        });

        // Bind add button
        this.$body.find(".lg-add-btn").on("click", () => this.add_item());

        // Bind load stock button
        this.$body.find(".lg-load-stock-btn").on("click", () => this.load_items_with_stock());

        // Bind load from receipt document button
        this.$body.find(".lg-load-receipt-btn").on("click", () => this.load_items_from_receipt_document());

        // Bind enter key on item field
        this.$body.find("#lg-item-field input").on("keypress", (e) => {
            if (e.which === 13) {
                this.add_item();
            }
        });
    }

    add_item() {
        const item_code = this.item_field.get_value();
        const qty = parseInt(this.qty_field.get_value()) || 1;

        if (!item_code) {
            frappe.show_alert({
                message: __("Please select an item"),
                indicator: "orange",
            });
            return;
        }

        // Check if item already in list
        const existing = this.items.find((i) => i.item_code === item_code);
        if (existing) {
            existing.qty += qty;
            this.render_items_list();
            this.item_field.set_value("");
            return;
        }

        // Get item details
        frappe.call({
            method:
                "retail_tools.retail_tools.page.label_generator.label_generator.get_item_for_label",
            args: { item_code },
            callback: (r) => {
                if (r.message && r.message.ok) {
                    this.items.push({
                        ...r.message.item,
                        qty: qty,
                    });
                    this.render_items_list();
                    this.item_field.set_value("");
                    this.qty_field.set_value(1);
                } else {
                    frappe.show_alert({
                        message: r.message?.message || __("Item not found"),
                        indicator: "red",
                    });
                }
            },
        });
    }

    render_items_list() {
        const $body = this.$body.find(".lg-items-body");
        const $empty = this.$body.find(".lg-items-empty");

        if (this.items.length === 0) {
            $body.html("");
            $empty.show();
            return;
        }

        $empty.hide();

        let html = "";
        this.items.forEach((item, idx) => {
            html += `
        <div class="lg-item-row" data-idx="${idx}">
          <span class="lg-col-item">${frappe.utils.escape_html(item.item_name)}</span>
          <span class="lg-col-barcode">${item.barcode || "-"}</span>
          <span class="lg-col-qty">
            <input type="number" class="form-control lg-qty-input" value="${item.qty}" min="1">
          </span>
          <span class="lg-col-actions">
            <button class="btn btn-xs btn-danger lg-remove-btn">
              <i class="fa fa-times"></i>
            </button>
          </span>
        </div>
      `;
        });

        $body.html(html);

        // Bind qty change
        $body.find(".lg-qty-input").on("change", (e) => {
            const idx = $(e.target).closest(".lg-item-row").data("idx");
            this.items[idx].qty = parseInt($(e.target).val()) || 1;
        });

        // Bind remove
        $body.find(".lg-remove-btn").on("click", (e) => {
            const idx = $(e.target).closest(".lg-item-row").data("idx");
            this.items.splice(idx, 1);
            this.render_items_list();
        });
    }

    clear_items() {
        this.items = [];
        this.render_items_list();
        this.$body.find(".lg-preview").html("");
    }

    load_items_with_stock() {
        const warehouse = this.warehouse_field.get_value();

        frappe.call({
            method:
                "retail_tools.retail_tools.page.label_generator.label_generator.get_items_with_stock",
            args: { warehouse: warehouse || "" },
            freeze: true,
            freeze_message: __("Loading items with stock..."),
            callback: (r) => {
                if (r.message && r.message.ok) {
                    // Merge with existing items
                    r.message.items.forEach((item) => {
                        const existing = this.items.find(
                            (i) => i.item_code === item.item_code
                        );
                        if (existing) {
                            existing.qty += item.qty;
                        } else {
                            this.items.push(item);
                        }
                    });
                    this.render_items_list();
                    frappe.show_alert({
                        message: __("{0} items loaded", [r.message.count]),
                        indicator: "green",
                    });
                } else {
                    frappe.show_alert({
                        message: r.message?.message || __("No items found"),
                        indicator: "orange",
                    });
                }
            },
        });
    }

    load_items_from_receipt_document() {
        const doctype = this.receipt_doctype_field.get_value();
        const docname = this.receipt_document_field.get_value();

        if (!docname) {
            frappe.show_alert({
                message: __("Please select a document"),
                indicator: "orange",
            });
            return;
        }

        frappe.call({
            method:
                "retail_tools.retail_tools.page.label_generator.label_generator.get_items_from_receipt_document",
            args: { doctype: doctype, docname: docname },
            freeze: true,
            freeze_message: __("Loading items from document..."),
            callback: (r) => {
                if (r.message && r.message.ok) {
                    // Merge with existing items
                    r.message.items.forEach((item) => {
                        const existing = this.items.find(
                            (i) => i.item_code === item.item_code
                        );
                        if (existing) {
                            existing.qty += item.qty;
                        } else {
                            this.items.push(item);
                        }
                    });
                    this.render_items_list();
                    frappe.show_alert({
                        message: __("{0} items loaded", [r.message.count]),
                        indicator: "green",
                    });
                } else {
                    frappe.show_alert({
                        message: r.message?.message || __("No items found"),
                        indicator: "orange",
                    });
                }
            },
        });
    }

    generate_labels() {
        if (this.items.length === 0) {
            frappe.show_alert({
                message: __("Please add items first"),
                indicator: "orange",
            });
            return;
        }

        const label_format = this.format_field.get_value() || "medium";
        const show_price = this.$body.find("#lg-show-price").is(":checked") ? 1 : 0;
        const price_list = this.price_list_field.get_value() || "";

        frappe.call({
            method:
                "retail_tools.retail_tools.page.label_generator.label_generator.generate_labels_html",
            args: {
                items: JSON.stringify(this.items),
                label_format: label_format,
                show_price: show_price,
                price_list: price_list,
            },
            freeze: true,
            freeze_message: __("Generating labels..."),
            callback: (r) => {
                if (r.message && r.message.ok) {
                    this.$body.find(".lg-preview").html(r.message.html);
                    this.init_barcodes();
                    frappe.show_alert({
                        message: __("{0} labels generated", [r.message.count]),
                        indicator: "green",
                    });
                } else {
                    frappe.show_alert({
                        message: r.message?.message || __("Error generating labels"),
                        indicator: "red",
                    });
                }
            },
        });
    }

    init_barcodes() {
        if (typeof JsBarcode === "undefined") {
            const script = document.createElement("script");
            script.src =
                "https://cdn.jsdelivr.net/npm/jsbarcode@3.11.5/dist/JsBarcode.all.min.js";
            script.onload = () => this.render_barcodes();
            document.head.appendChild(script);
        } else {
            this.render_barcodes();
        }
    }

    render_barcodes() {
        const format = this.format_field.get_value();
        let height = 30;
        if (format === "large") height = 55;
        if (format === "small") height = 25;

        this.$body.find(".barcode-svg").each(function () {
            const barcode = $(this).data("barcode");
            if (barcode) {
                try {
                    JsBarcode(this, String(barcode), {
                        format: "CODE128",
                        width: 1.2,
                        height: height,
                        displayValue: true,
                        fontSize: 10,
                        margin: 2,
                    });
                } catch (e) {
                    console.log("Barcode error:", e);
                }
            }
        });
    }

    print_labels() {
        const $preview = this.$body.find(".lg-preview");
        if (!$preview.html()) {
            frappe.show_alert({
                message: __("Please generate labels first"),
                indicator: "orange",
            });
            return;
        }

        const printContent = $preview.html();
        const printWindow = window.open("", "_blank");
        printWindow.document.write(`
      <!DOCTYPE html>
      <html>
      <head>
        <title>${__("Barcode Labels")}</title>
        <style>
          body { margin: 0; padding: 10px; }
          @media print {
            body { padding: 0; }
          }
        </style>
      </head>
      <body>
        ${printContent}
        <script>window.print(); window.close();</script>
      </body>
      </html>
    `);
        printWindow.document.close();
    }
};
